src/main/java/algorithm/VisualizationManager.java

```diff
@@ -0,0 +1,395 @@
+
+        // loads the next set of tracking data
+        trackingService.getTrackingDataSource().update();
+        List<ToolMeasure> tools = trackingService.getDataService().loadNextData(1);
+
+        int j = 0;
+
+        for (ToolMeasure tool : tools) {
+            List<Measurement> measurements = tool.getMeasurement();
+            float[] eulerAngles = new float[3];
+
+            Quaternion rotationMovement = tool.getMeasurement().get(measurements.size() - 1).getRotation();
+            rotationMovement.toAngles(eulerAngles);
+
+            double x = tool.getMeasurement().get(measurements.size() - 1).getPoint().getX();
+            // invert Y & Z axis for correct display!
+            double y = tool.getMeasurement().get(measurements.size() - 1).getPoint().getY() * -1;
+            double z = tool.getMeasurement().get(measurements.size() - 1).getPoint().getZ() * -1;
+            // convert Quaternion to Euler
+            float yaw = eulerAngles[0];
+            float roll = eulerAngles[1];
+            float pitch = eulerAngles[2];
+            // convert Euler to angles
+            double yawAngle = Math.toDegrees(yaw);
+            double rollAngle = Math.toDegrees(roll);
+            double pitchAngle = Math.toDegrees(pitch);
+
+            boolean showRotations = false;
+            if (showRotations) {
+                System.out.println(tool.getName());
+                System.out.println("yaw:!" + yaw);
+                System.out.println("angle!" + yawAngle);
+                System.out.println("roll:!" + roll);
+                System.out.println("angle!" + rollAngle);
+                System.out.println("pitch:!" + pitch);
+                System.out.println("angle!" + pitchAngle);
+                System.out.print("\n");
+            }
+
+            // apply translation and rotation to each tracker
+            trackingCones[j].getTransforms().setAll(
+                    new Translate(x, y, z),
+                    new Rotate(Math.toDegrees(yaw), Rotate.Y_AXIS),
+                    new Rotate(Math.toDegrees(pitch), Rotate.X_AXIS),
+                    new Rotate(Math.toDegrees(roll), Rotate.Z_AXIS)
+            );
+
+            checkBounds();
+            if (j < tools.size()) {
+                j++;
+            }
+        }
+    }
+
+    /**
+     * Checks for collision and changes tracker color accordingly
+     * Color red implies a collision with the model was detected
+     * Color green implies no collision was detected
+     */
+    private void checkBounds() {
+        boolean collisionDetected = false;
+        for (ConeMesh trackingCone : trackingCones) {
+            trackingCone.setMaterial(new PhongMaterial(Color.GREEN));
+
+            for (MeshView stlFile : stlFiles
+            ) {
+
+                if (trackingCone.getBoundsInParent().intersects(stlFile.getBoundsInParent())) {
+                    collisionDetected = true;
+                }
+
+                if (collisionDetected) {
```

Comment:
You could simplify this check: directly color the cone red when a collision is detected